src/common/admin_socket.cc

```diff
@@ -506,7 +506,28 @@ void AdminSocket::execute_command(
 		     empty);
   }
 
-  auto f = Formatter::create(format, "json-pretty", "json-pretty");
+  string output;
+  try {
+    cmd_getval(cmdmap, "output-file", output);
+  } catch (const bad_cmd_get& e) {
+    output = "";
+  }
+
+  Formatter* f;
+  if (format == "json-file") {
+    if (output.size()) {
+      auto* jff = new JSONFormatterFile(output, false);
+      auto&& of = jff->get_ofstream();
```

Comment:
`auto&& var` is a deducting reference type, which can effectively be l-value or r-value depending on the initializer. See https://en.cppreference.com/w/cpp/language/auto#Explanation:

> For example, given `const auto& i = expr;`, the type of i is exactly the type of the argument u in an imaginary `template template<class U> void f(const U& u)` if the function call `f(expr)` was compiled. Therefore, `auto&&` may be deduced either as an lvalue reference or rvalue reference according to the initializer, which is used in range-based for loop.

Reply:
Yeah, that a definition. Still: what's the benefit behind the extra visual clutter? Later it's used as regular l-value reference. Do we get anything from increasing mental load on humans who read this?

Reply:
I'd promote the usage of `auto &&` so that people would be less intimidated by it. 
However, I agree that in this context this isn't useful. If anything, I'd suggest an `auto const&` here.

Reply:
For what it's worth, my thoughts on using `auto&&` is that it makes the caller more resilient to possible minor type changes like a  `T*` to `std::shared_ptr<T>`. If we had:

`auto* T = foo();`

then that definition would need updated. Similarly:

`auto T = foo();`

is invalid if `T` changes from a pointer to `std::unique_ptr<T>&`.

On the other hand,

`auto&& T = foo();`

is the most generic and so long as `T` still behaves like a pointer, we don't care how it's wrapped/proxied. (Similar logic can be constructed for why `auto&&` is preferable even if presently the return value of `foo()`  is a `T&` as is the case here.)

Reply:
The downside of using `auto&&` in this context is that it takes the cv-qualification of the initializer expression, while, the constness of this local variable should be controlled by the call site rather than by the API. In this block, we want a const reference to whatever is returned, regardless of whether the initializer provides a const object. 

This is in contrast with the use case of

```
for(auto&& i = std::cbegin(vec); ...)
```
, where we still control the constness at the call site, but with the initializer expression rather than the type definition.

Reply:
@batrick you encouraged me for the following test. The TL;DR; is that `auto &&` is a shorter form of `auto const&` as far as the call-site is concerned, so its usage is justified 👍🏻.
The potential unsafe usage of a mutable reference to a managed ptr is more of an API issue.

```
#include <vector>
#include <memory>

struct C {
    std::vector<int> items;
};

std::shared_ptr<C> global_c(new C());
std::unique_ptr<C> global_unique_c(new C());

C*                          get_p()     { return global_c.get(); }
C const*                    get_cp()    { return global_c.get(); }
std::shared_ptr<C>          get_s()     { return global_c; }
std::shared_ptr<C> &        get_rs()    { return global_c; }
std::shared_ptr<C> const&   get_crs()   { return global_c; }
std::unique_ptr<C>          get_u()     { return std::make_unique<C>(); }
std::unique_ptr<C> &        get_ru()    { return global_unique_c; }
std::unique_ptr<C> const&   get_cru()   { return global_unique_c; }

int main(int argc, char** argv)
{
    { auto c = get_p(); c->items.size(); }
    { auto c = get_cp(); c->items.size(); }
    { auto c = get_s(); c->items.size(); }
    { auto c = get_rs(); c->items.size(); }
    { auto c = get_crs(); c->items.size(); }
    { auto c = get_u(); c->items.size(); }
    // { auto c = get_ru(); c->items.size(); }     // error: call to implicitly-deleted copy constructor of 'std::unique_ptr<C>'
    // { auto c = get_cru(); c->items.size(); }    // error: call to implicitly-deleted copy constructor of 'std::unique_ptr<C>'

    // { auto& c = get_p(); c->items.size(); }     // error: non-const lvalue reference cannot bind to a temporary
    // { auto& c = get_cp(); c->items.size(); }    // error: non-const lvalue reference cannot bind to a temporary
    // { auto& c = get_s(); c->items.size(); }     // error: non-const lvalue reference cannot bind to a temporary
    { auto& c = get_rs(); c->items.size(); }
    { auto& c = get_crs(); c->items.size(); }
    // { auto& c = get_u(); c->items.size(); }     // error: non-const lvalue reference cannot bind to a temporary
    { auto& c = get_ru(); c->items.size(); }
    { auto& c = get_cru(); c->items.size(); }


    { auto&& c = get_p(); c->items.size(); }
    { auto&& c = get_cp(); c->items.size(); }
    { auto&& c = get_s(); c->items.size(); }
    { auto&& c = get_rs(); c->items.size(); }
    { auto&& c = get_crs(); c->items.size(); }
    { auto&& c = get_u(); c->items.size(); }
    { auto&& c = get_ru(); c->items.size(); }
    { auto&& c = get_cru(); c->items.size(); }

    { auto const& c = get_p(); c->items.size(); }
    { auto const& c = get_cp(); c->items.size(); }
    { auto const& c = get_s(); c->items.size(); }
    { auto const& c = get_rs(); c->items.size(); }
    { auto const& c = get_crs(); c->items.size(); }
    { auto const& c = get_u(); c->items.size(); }
    { auto const& c = get_ru(); c->items.size(); }
    { auto const& c = get_cru(); c->items.size(); }

    {
        auto&& c = get_ru();
        // this is valid code, but it's unlikely we ever want a mutable reference to a managed ptr
        // however, IMO it's more of a problem with the API
        [&] { c.reset(); }(); // unhandled race if this is executed by a separate thread
    }

    {
        auto const& c = get_ru();
        // [&] { c.reset(); }(); // error: 'this' argument to member function 'reset' has type 'const std::unique_ptr<C>', but function is not marked const
    }
}
```